### PR TITLE
192 feature refurbish decision tree intersection

### LINKED ISSUE #192: [Feature]: refurbish decision tree - intersection
### Description

Description

Continuing to adapt the old decision tree solution of the old team to work with our agent.

This will be a longer lasting part of the project, so it will be broken down in different parts.

The next logical part is gonna be the behavior around intersections as it probably makes up a large part of the scenarios we are gonna encounter

### Definition of Done

Create a couple of dummy topics to test if the right parts of the decision tree are reached and the decision tree continues working correctly.

### PR DESCRIPTION
**Pull request by JoKircher**
# Description

This PR reworked the most part of the intersection handling logic. A mock publisher was created to simulate a traffic light. If a red or yellow light is detected, the vehicle stay in the wait behavior until a green light is detected. For now, if there is green light, the subtree repeats itself as the code for leaving the intersection is still missing/in a draft PR #197. As most of the functionality is there and the rest of the PRs will hopefully shortly be able to be merged I decided to mark this as a PR, not a draft

Fixes #192 

## Time invested

Josef Kircher 8 h

## Type of change

Please delete options that are not relevant.

- New feature (non-breaking change which adds functionality)

## Does this PR introduce a breaking change?

no

## Most important changes

intersection.py

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works (might be obsolete with CI later on)
- [x] New and existing unit tests pass locally with my changes (might be obsolete with CI later on)

### CODE REVIEW

**Review by schwalga — CHANGES_REQUESTED** [review]
Very nice PR. Looks like you made a lot of progress. 
Your presentation definitely helped in understanding the desired behavior. 
My only small suggestion would be to either create a method for the conversion between kph and ms as the current way doesn't seem super elegant, or directly use ms values in your code.
Apart from that, I think this PR is ready to be merged, with the next step being testing with the global planner. 
Thanks as always for your great documentation, which helps a lot when reviewing in an unknown domain.

File: code/mock/src/mock_traffic_light.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+from ros_compatibility.node import CompatibleNode
+from rospy import Publisher
+# from std_msgs.msg import Float32
+from mock.msg import Traffic_light
+
+
+class MockTrafficLightPublisher(CompatibleNode):
+    """
+    This node publishes traffic light information. It can be used for testing.
+    TODO functionality
```
[the remark this answers is not in the document]
Comment (JoKircher) [reply code/mock/src/mock_traffic_light.py:12]: No, I will change this.

File: code/test_planning_top/behavior_agent/src/behavior_agent/behaviours/intersection.py [hunk 2]
```diff
@@ -75,35 +122,36 @@ def update(self):
...
+        elif speed > 5.0/KMH_TO_MS and self.virtual_stopline_distance < 3.5:
             # running over line
             return py_trees.common.Status.SUCCESS
         elif self.last_virtual_distance == self.virtual_stopline_distance \
-                and self.virtual_stopline_distance < 10:
+                and self.virtual_stopline_distance < 10.0:
             # ran over line
             return py_trees.common.Status.SUCCESS
 
         next_lanelet_msg = self.blackboard.get("/psaf/ego_vehicle/"
```
[the remark this answers is not in the document]
Comment (JoKircher) [reply code/test_planning_top/behavior_agent/src/behavior_agent/behaviours/intersection.py:152]: Yes, this needs to be addressed once the global route is available and testing is facilitated.

File: code/test_planning_top/behavior_agent/src/behavior_agent/behaviours/intersection.py [hunk 3]
```diff
@@ -35,20 +68,33 @@ def initialise(self):
         self.traffic_light_distance = np.inf
         self.traffic_light_status = ''
         self.virtual_stopline_distance = np.inf
-        self.target_speed_pub.publish(30.0)
+        self.target_speed_pub.publish(30.0/KMH_TO_MS)
```
[the remark this answers is not in the document]
Comment (JoKircher) [reply code/test_planning_top/behavior_agent/src/behavior_agent/behaviours/intersection.py]: fixed in https://github.com/ll7/paf22/pull/198/commits/fc915c6c60557e86089f9bd8a7200c6b6f64d1bc

File: code/mock/src/mock_traffic_light.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (JoKircher) [reply code/mock/src/mock_traffic_light.py]: https://github.com/ll7/paf22/pull/198/commits/fc915c6c60557e86089f9bd8a7200c6b6f64d1bc

File: code/mock/src/mock_traffic_light.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        self.new_timer(self.control_loop_rate, loop)
+        self.spin()
+
+
+def main(args=None):
+    """
+    Main function starts the node
+    :param args:
+    """
+    roscomp.init('velocity_publisher_dummy', args=args)
```
Comment (schwalga) [inline code/mock/src/mock_traffic_light.py:66]: Sorry, i missed this on the first review, but I think it should be traffic_light_dummy.

**Review by schwalga — CHANGES_REQUESTED** [review]
Thanks for implementing the feedback. I found one more minor issue, after which the PR is ready to be merged.

File: code/mock/src/mock_traffic_light.py (hunk 4 above)
Comment (JoKircher) [reply code/mock/src/mock_traffic_light.py]: fixed in https://github.com/ll7/paf22/pull/198/commits/37e4b8b9587048eb1bad97e3ff76001718a0acb8

**Review by schwalga — APPROVED** [review]
Thanks for implementing the feedback. Great PR.

**Pull request merged**